packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -63,8 +63,62 @@ async function findAllHotswappableChanges(
   let foundNonHotswappableChange = false;
   const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
 
+  // as a first pass, we need to collapse replacement changes into one,
+  // as they are represented here as a pair of two changes: one addition and one removal
+  const allResourceChanges: { [logId: string]: cfn_diff.ResourceDifference } = stackChanges.resources.changes;
+  const allRemovalChanges = Object.entries(allResourceChanges).reduce((acc, [logId, resChange]) => {
+    if (resChange.isRemoval) { acc[logId] = resChange; }
+    return acc;
+  }, {} as { [logId: string]: cfn_diff.ResourceDifference });
+  const allNonRemovalChanges = Object.entries(allResourceChanges).reduce((acc, [logId, resChange]) => {
+    if (!resChange.isRemoval) { acc[logId] = resChange; }
+    return acc;
+  }, {} as { [logId: string]: cfn_diff.ResourceDifference });
+  const allResourceChangesWithReplacements: { [logId: string]: cfn_diff.ResourceDifference } = {};
+  for (const [logId, nonRemovalChange] of Object.entries(allNonRemovalChanges)) {
+    if (nonRemovalChange.isAddition) {
+      const addChange = nonRemovalChange;
+      // search for an identical removal change
+      const identicalRemovalChange = Object.entries(allRemovalChanges).find(([_, remChange]) => {
```

Comment:
Yep.

(We can't do it there, because `find()` can return `undefined`, but I did it in an `if` that checks whether `identicalRemovalChange` is non-`undefined`)